### PR TITLE
feat: apply exponential backoff on worker restarts

### PR DESCRIPTION
**Pull request by pmalouin**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -32,6 +32,13 @@ function getVersion () {
 
 var version = getVersion();
 
+function getRestartDelay (consecutive_crashes_count) {
+  if (process.env.RESTART_BACKOFF === 'constant') {
+    return WORKER_THROTTLE;
+  }
+  return WORKER_THROTTLE * Math.pow(2, consecutive_crashes_count - 1);
```
Comment (luuuis) [inline index.js:39]: There needs to be a ceiling or you could very quickly start scheduling things into next week.

**Review by luuuis — COMMENTED** [review]
I don't think exponential backoff on a per-worker basis is a good fit. Exponential backoff is used as a way for **independent** processes to collaborate: usually reducing contention on a given resource and to achieve max throughput in the face of retries. In this case the processes are not independent: master-process is in control of all of them and can coordinate the entire thing.

To _safely_ achieve the end result of not spawning workers too quickly it would make more sense to throttle worker spawn on a cluster basis rather than per-worker. Essentially setting a maximum rate of starting workers.

The reason I did not do so initially is that it can lead to worker starvation if not done carefully. As can the current throttling strategy, but only for 1 second (a lot longer with this change).

File: index.js (hunk 1 above)
Comment (pmalouin) [reply index.js:39]: good point, thanks!

**Comment by pmalouin** [comment]
> I don't think exponential backoff on a per-worker basis is a good fit. Exponential backoff is used as a way for independent processes to collaborate: usually reducing contention on a given resource and to achieve max throughput in the face of retries. In this case the processes are not independent: master-process is in control of all of them and can coordinate the entire thing.
> To safely achieve the end result of not spawning workers too quickly it would make more sense to throttle worker spawn on a cluster basis rather than per-worker. Essentially setting a maximum rate of starting workers.
> The reason I did not do so initially is that it can lead to worker starvation if not done carefully. As can the current throttling strategy, but only for 1 second (a lot longer with this change).

I first went with a single/global crash counter for all workers and a naive exponential backoff with that counter, but I found an edge case that poses a problem. Let's say all workers crash at the same time. That would put the crash counter at N and we would already start to delay worker restart rates. My expectations for when all workers crash together is to try to re-spawn them quickly: we want to recover the capacity asap. If the re-spawns crash again, then we should start backing off.

So I agree with the suggestion to use a global coordination strategy, but IMO we need a smarter way to account for these kinds of cases. Maybe we can just introduce a non-linear back-off where the first N attempts are not delayed, then we start applying back-off? Thoughts or ideas?

**Comment by luuuis** [comment]
> My expectations for when all workers crash together is to try to re-spawn them quickly: we want to recover the capacity asap. If the re-spawns crash again, then we should start backing off.

Haven't you described a rate limiter that allows bursts of M and an otherwise steady rate of N/sec?

But you want to decrease the rate limit progressively over time, which doesn't make sense for this problem: there is no benefit in trending towards 0% CPU usage as there are no other participants contending on that CPU other than these workers. This is in contrast to typical client-server scenarios where there is no such upper bound.

> we need a smarter way to account for these kinds of cases

The smartest thing for master-process to do is to respawn workers as quickly as possible without overwhelming CPU and the smartest thing for the workers is to do is to not crash when encountering transient errors. :)

**Comment by pmalouin** [comment]
Appreciate the thoughtful feedback @luuuis 🙇. My primary aim was less about CPU usage of the service itself, but rather the cascading effect it can have on other back-ends (ex. repeatedly opening db connections).

Although at this point, I'm convinced to close this PR as our current `WORKER_THROTTLE` meets the requirements of the constant rate-limit with an initial burst. Some manual tweaking could be warranted on some of our services to increase the value as we see fit (higher than the default 1s), to throttle more aggressively restarts. But there is no need to progressively slow them down 👍

**Pull request closed without merging**